engines/wage/world.cpp

```diff
@@ -0,0 +1,510 @@
+			Common::String message = readPascalString(res);
+			if (i == itemIndex) {
+				Common::String *msg = new Common::String(message);
+				delete res;
+				return msg;
+			}
+		}
+
+		delete res;
+	}
+
+	return NULL;
+}
+
+bool InvComparator(Obj *l, Obj *r) {
+    return l->_index < r->_index;
+}
+
+void World::move(Obj *obj, Chr *chr) {
+	if (obj == NULL)
+		return;
+
+	Designed *from = obj->removeFromCharOrScene();
+	obj->_currentOwner = chr;
+	chr->_inventory.push_back(obj);
+
+	Common::sort(chr->_inventory.begin(), chr->_inventory.end(), InvComparator);
+
+	_engine->onMove(obj, from, chr);
+}
+
+bool ObjComparator(Obj *o1, Obj *o2) {
+	bool o1Immobile = (o1->_type == Obj::IMMOBILE_OBJECT);
+	bool o2Immobile = (o2->_type == Obj::IMMOBILE_OBJECT);
+	if (o1Immobile == o2Immobile) {
+		return o1->_index - o2->_index;
+	}
+	return o1Immobile;
+}
+
+void World::move(Obj *obj, Scene *scene, bool skipSort) {
+	if (obj == NULL)
+		return;
+
+	Designed *from = obj->removeFromCharOrScene();
+	obj->_currentScene = scene;
+	scene->_objs.push_back(obj);
+
+	if (!skipSort)
+		Common::sort(scene->_objs.begin(), scene->_objs.end(), ObjComparator);
+
+	_engine->onMove(obj, from, scene);
+}
+
+bool ChrComparator(Chr *l, Chr *r) {
+    return l->_index < r->_index;
+}
+
+void World::move(Chr *chr, Scene *scene, bool skipSort) {
+	if (chr == NULL)
+		return;
+	Scene *from = chr->_currentScene;
+	if (from == scene)
+		return;
+	if (from != NULL)
+		from->_chrs.remove(chr);
+	scene->_chrs.push_back(chr);
+
+	if (!skipSort)
+		Common::sort(scene->_chrs.begin(), scene->_chrs.end(), ChrComparator);
+
+	if (scene == _storageScene) {
+		chr->resetState();
+	} else if (chr->_playerCharacter) {
+		scene->_visited = true;
+		_player->_context._visits++;
+	}
+	chr->_currentScene = scene;
+
+	_engine->onMove(chr, from, scene);
+}
+
+Scene *World::getRandomScene() {
+	// Not including storage:
+	return _orderedScenes[1 + _engine->_rnd->getRandomNumber(_orderedScenes.size() - 2)];
+}
+
+Scene *World::getSceneAt(int x, int y) {
+	for (uint i = 0; i < _orderedScenes.size(); i++) {
+		Scene *scene = _orderedScenes[i];
+
+		if (scene != _storageScene && scene->_worldX == x && scene->_worldY == y) {
+			return scene;
+		}
+	}
+	return NULL;
+}
+
+static const int directionsX[] = { 0, 0, 1, -1 };
+static const int directionsY[] = { -1, 1, 0, 0 };
+
+bool World::scenesAreConnected(Scene *scene1, Scene *scene2) {
+	if (!scene1 || !scene2)
+		return false;
+
+	int x = scene2->_worldX;
+	int y = scene2->_worldY;
+
+	for (int dir = 0; dir < 4; dir++)
+		if (!scene2->_blocked[dir])
+			if (getSceneAt(x + directionsX[dir], y + directionsY[dir]) == scene1)
+				return true;
+
+	return false;
+}
+
+const char *World::getAboutMenuItemName() {
+	static char menu[256];
+
+	*menu = '\0';
+
+	if (_aboutMenuItemName.size() == 0) {
+		sprintf(menu, "About %s...", _name.c_str());
+	} else { // Replace '@' with name
+		const char *str = _aboutMenuItemName.c_str();
+		char *pos = strchr(str, '@');
```

Comment:
We need to use `const char *` for `pos` here. Otherwise the compiler will try to pick the version of `strchr` taking `char *` resulting in a compilation error.